### PR TITLE
Add UserNotifications Framework for iOS 10+

### PR DESCRIPTION
**Pull request by mrgrauel**
I added a permission for the new notification framework UserNotifications in iOS 10 to request permission with the new api.

* add permission `UserNotification`
* add configuration `PERMISSION_USERNOTIFICATIONS`
* add permissionType `userNotifications`
* removed the file private static variable `_notification` cause it's never read
* did **not** change the `requestedNotifications` default cause it returns the same information

---

I tried to keep it in the structure you designed the framework.

### CODE REVIEW

**Comment by mrgrauel** [comment]
On my local machine the tests do run. Does anybody has an idea why the checks fail?

**Comment by mrgrauel** [comment]
@delba what's going on?

**Comment by pasevin** [comment]
Any chance this is going to be merged?

File: Source/Permission.swift [hunk 1]
```diff
@@ -98,31 +116,22 @@ open class Permission: NSObject {
         return Permission(type: .notifications(settings))
     }()
     
-    /// Variable used to retain the notifications permission.
-    fileprivate static var _notifications: Permission?
```
Comment (delba) [inline Source/Permission.swift]: This variable is used to retain the permission as stated in the comment above.

File: Source/Permission.swift [hunk 2]
```diff
@@ -98,31 +116,22 @@ open class Permission: NSObject {
...
-    
     /// The permission to send notifications.
     open static func notifications(types: UIUserNotificationType, categories: Set<UIUserNotificationCategory>?) -> Permission {
-        let settings   = UIUserNotificationSettings(types: types, categories: categories)
-        let permission = Permission(type: .notifications(settings))
-        _notifications = permission
```
Comment (delba) [inline Source/Permission.swift]: `_notifications` is used to retain the permission

File: Source/Permission.swift [hunk 3]
```diff
@@ -98,31 +116,22 @@ open class Permission: NSObject {
...
-        return permission
+        let settings = UIUserNotificationSettings(types: types, categories: categories)
+        return Permission(type: .notifications(settings))
     }
     
     /// The permission to send notifications.
     open static func notifications(types: UIUserNotificationType) -> Permission {
         let settings   = UIUserNotificationSettings(types: types, categories: nil)
-        let permission = Permission(type: .notifications(settings))
-        _notifications = permission
```
Comment (delba) [inline Source/Permission.swift:122]: Same comment as above

File: Source/Permission.swift [hunk 4]
```diff
@@ -98,31 +116,22 @@ open class Permission: NSObject {
...
-        return permission
+        return Permission(type: .notifications(settings))
     }
     
     /// The permission to send notifications.
     open static func notifications(categories: Set<UIUserNotificationCategory>?) -> Permission {
         let settings  = UIUserNotificationSettings(types: [.badge, .sound, .alert], categories: categories)
-        let permission = Permission(type: .notifications(settings))
-        _notifications = permission
```
Comment (delba) [inline Source/Permission.swift:130]: Again

File: Source/Permission.swift [hunk 5]
```diff
@@ -90,6 +93,21 @@ open class Permission: NSObject {
...
+    @available(iOS 10.0, *)
+    open static let userNotifications: Permission = {
+        let settings: UNAuthorizationOptions = [.alert, .badge, .sound]
+        return Permission(type: .userNotifications(settings))
+    }()
+    
+    /// The permission to send notifications.
+    @available(iOS 10.0, *)
+    open static func userNotifications(options: UNAuthorizationOptions) -> Permission {
+        return Permission(type: .userNotifications(options))
```
Comment (delba) [inline Source/Permission.swift]: We need to retain the permission with a private var like in the case of `notifications` (e.g: `_userNotifications`)

File: Source/PermissionType.swift [hunk 6]
```diff
@@ -21,6 +21,9 @@
 // OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 // SOFTWARE.
 //
+#if PERMISSION_NOTIFICATIONS
```
Comment (delba) [inline Source/PermissionType.swift]: Should be `PERMISSION_USERNOTIFICATIONS`

File: Permission.podspec [hunk 7]
```diff
@@ -1,6 +1,6 @@
 Pod::Spec.new do |s|
   s.name         = "Permission"
-  s.version      = "2.0.3"
+  s.version      = "2.1.0"
```
Comment (delba) [inline Permission.podspec]: I prefer to change it just before pushing it (and it's dependent on the git tag)

File: Permission.podspec [hunk 8]
```diff
@@ -64,6 +64,11 @@ Pod::Spec.new do |s|
     no.pod_target_xcconfig = { "SWIFT_ACTIVE_COMPILATION_CONDITIONS"  => "PERMISSION_NOTIFICATIONS" }
   end
 
+  s.subspec 'UserNotifications' do |no|
+    no.dependency 'Permission/Core'
+    no.pod_target_xcconfig = { "SWIFT_ACTIVE_COMPILATION_CONDITIONS"  => "PERMISSION_USERNOTIFICATIONS" }
```
Comment (delba) [inline Permission.podspec]: Could we rename `PERMISSION_USERNOTIFICATIONS` to `PERMISSION_USER_NOTIFICATIONS` for the sake of consistency please?

File: Source/PermissionTypes/UserNotifications.swift [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+//
+//  Created by Andreas Grauel on 20.01.17.
+//  Copyright © 2017 delba. All rights reserved.
+//
+
+#if PERMISSION_USERNOTIFICATIONS
+import UserNotifications
+
+internal extension Permission {
+    var statusUserNotifications: PermissionStatus {
```
Comment (delba) [inline Source/PermissionTypes/UserNotifications.swift]: Could we keep the structure of the `Motion` permission (that also uses "synchronous" calls) please ?

**Review by delba — CHANGES_REQUESTED** [review]

**Comment by delba** [comment]
Hi @mrgrauel, thanks a lot for the PR! There are a just few changes to do before merging though.
Thanks!

**Comment by mrgrauel** [comment]
Good Morning @delba, is it now as you want it?

File: Source/Permission.swift (hunk 2 above)
Comment (mrosberghaus) [reply Source/Permission.swift]: Changed

File: Source/Permission.swift (hunk 3 above)
Comment (mrosberghaus) [reply Source/Permission.swift:122]: Changed

File: Source/Permission.swift (hunk 4 above)
Comment (mrosberghaus) [reply Source/Permission.swift:130]: Changed

File: Source/Permission.swift (hunk 2 above)
Comment (mrosberghaus) [reply Source/Permission.swift]: changed

**Review by mrgrauel — COMMENTED** [review]
Changed

**Comment by mrgrauel** [comment]
@delba I changed everything. I dunno why it still says that a change request is open.

**Comment by mrgrauel** [comment]
@delba Hello?!?!?!

**Comment by otymartin** [comment]
@delba Has this issue been fixed? I still get these xcode 10 deprecation warnings. It wont compile with target iOS 10

**Review by rob5408 — APPROVED** [review]

**Comment by rob5408** [comment]
^^ @delba Sorry about that, I thought I was on my fork. I know it doesn't actually merge the PR, but why does GitHub even give me the option. :\

**Comment by zhuorantan** [comment]
@delba Please merge this PR

**Comment by pasevin** [comment]
Losing hope :/

**Comment by Jeehut** [comment]
Hey everybody, I've **given up on waiting** for @delba to answer and merge our PRs. Instead I've **integrated** this PR into my fork (which I created for #94). I've also created my own CocoaPods entry so we can continue using this framework and merging our improvements together. Feel free to star and send your PRs to my fork in the future. I will try to review them from time to time:
https://github.com/Dschee/Permission

To switch to my fork in your project simply replace `pod 'Permission'` with `pod 'Dschee-Permission'` if you are using CocoaPods or `github "delba/Permission"` with `github "Dschee/Permission"` if your are using Carthage. I've also made a new release: **v2.1.0**.

**Comment by otymartin** [comment]
@Dschee nice move 👍

**Comment by delba** [comment]
Hi @mrgrauel, the project has been updated to the latest Swift version and now uses `UserNotifications` (https://github.com/delba/Permission/blob/master/Source/Types/Notifications.swift).

**Pull request closed without merging**